### PR TITLE
Add Event for Engine

### PR DESCRIPTION
**Pull request by yaooqinn**
### _How was this patch tested?_
- [x] Add some test cases that check the changes thoroughly including negative and positive cases if possible

- [ ] Add screenshots for manual tests if appropriate

- [x] [Run test](https://kyuubi.readthedocs.io/en/latest/tools/testing.html#running-tests) locally before make a pull request

### CODE REVIEW

File: externals/kyuubi-spark-monitor/pom.xml [hunk 1]
```diff
@@ -32,6 +32,12 @@
     <name>Kyuubi Project Spark Monitor</name>
 
     <dependencies>
+
+        <dependency>
+            <groupId>org.scala-lang</groupId>
+            <artifactId>scala-library</artifactId>
```
Comment (pan3793) [inline externals/kyuubi-spark-monitor/pom.xml:38]: `kyuubi-common` depends on `scala-library`, seems no require to declare again?

**Comment by yaooqinn** [comment]
cc @zhang1002 @ulysses-you

File: externals/kyuubi-spark-monitor/pom.xml (hunk 1 above)
Comment (yaooqinn) [reply externals/kyuubi-spark-monitor/pom.xml]: removed

File: externals/kyuubi-spark-sql-engine/src/main/scala/org/apache/kyuubi/engine/spark/events/EngineEvent.scala [hunk 2]
```diff
@@ -0,0 +1,147 @@
...
+ * @param endTime end time
+ * @param state the engine state
+ * @param diagnostic caught exceptions if any
+ */
+case class EngineEvent(
+    applicationId: String,
+    attemptId: Option[String],
+    applicationName: String,
+    owner: String,
+    shareLevel: String,
```
Comment (ulysses-you) [inline externals/kyuubi-spark-sql-engine/src/main/scala/org/apache/kyuubi/engine/spark/events/EngineEvent.scala:56]: shall we record `subDomain` ?

File: externals/kyuubi-spark-sql-engine/src/main/scala/org/apache/kyuubi/engine/spark/events/EngineEvent.scala [hunk 3]
```diff
@@ -0,0 +1,147 @@
...
+    deployMode: String,
+    sparkVersion: String,
+    webUrl: String,
+    driverCores: Int,
+    driverMemoryMB: Int,
+    executorCores: Int,
+    executorMemoryMB: Int,
+    maxExecutors: Int,
+    startTime: Long,
+    var endTime: Long = -1L,
```
Comment (ulysses-you) [inline externals/kyuubi-spark-sql-engine/src/main/scala/org/apache/kyuubi/engine/spark/events/EngineEvent.scala:68]: if we use `copy`, we don't need `var` anymore

File: externals/kyuubi-spark-sql-engine/src/main/scala/org/apache/kyuubi/engine/spark/events/EngineEvent.scala [hunk 4]
```diff
@@ -0,0 +1,147 @@
...
+    // need to consider deploy mode and cluster to get core and mem
+    val driverCores = sc.getConf.getInt("spark.driver.cores", 0)
+    val driverMemory = sc.getConf.getSizeAsMb("spark.driver.memory", "1g").toInt
+    val executorCore = sc.getConf.getInt("spark.executor.cores", 1)
+    val executorMemory = sc.getConf.getSizeAsMb("spark.executor.memory", "1g").toInt
+    val dae = sc.getConf.getBoolean("spark.dynamicAllocation.enabled", defaultValue = false)
+    val maxExecutors = if (dae) {
+      sc.getConf.getInt("spark.dynamicAllocation.maxExecutors", Int.MaxValue)
+    } else {
+      sc.getConf.getInt("spark.executor.instances", 1)
```
Comment (ulysses-you) [inline externals/kyuubi-spark-sql-engine/src/main/scala/org/apache/kyuubi/engine/spark/events/EngineEvent.scala:127]: the default value is 2

File: externals/kyuubi-spark-sql-engine/src/main/scala/org/apache/kyuubi/engine/spark/events/EngineEvent.scala [hunk 5]
```diff
@@ -0,0 +1,147 @@
...
+
+object EngineEvent {
+
+  def apply(engine: SparkSQLEngine): EngineEvent = {
+    val sc = engine.spark.sparkContext
+    val webUrl = sc.getConf.getOption(
+      "spark.org.apache.hadoop.yarn.server.webproxy.amfilter.AmIpFilter.param.PROXY_URI_BASES")
+      .orElse(sc.uiWebUrl).getOrElse("")
+    // need to consider deploy mode and cluster to get core and mem
+    val driverCores = sc.getConf.getInt("spark.driver.cores", 0)
```
Comment (ulysses-you) [inline externals/kyuubi-spark-sql-engine/src/main/scala/org/apache/kyuubi/engine/spark/events/EngineEvent.scala:119]: 1 ? 0 seems overkill

File: kyuubi-common/src/main/scala/org/apache/kyuubi/config/KyuubiConf.scala [hunk 6]
```diff
@@ -678,4 +678,28 @@ object KyuubiConf {
       .version("1.3.0")
       .booleanConf
       .createWithDefault(false)
+
+  val ENGINE_EVENT_JSON_LOG_PATH: ConfigEntry[String] =
+    buildConf("engine.event.json.log.path")
+      .doc("The location of all the engine events go for the builtin JSON logger")
+      .version("1.3.0")
+      .stringConf
+      .createWithDefault("/tmp/events")
```
Comment (ulysses-you) [inline kyuubi-common/src/main/scala/org/apache/kyuubi/config/KyuubiConf.scala:687]: `/tmp/kyuubi/events`

File: kyuubi-common/src/main/scala/org/apache/kyuubi/config/KyuubiConf.scala [hunk 7]
```diff
@@ -678,4 +678,28 @@ object KyuubiConf {
...
+        " <li>SPARK: the events will be written to the spark history events</li>" +
+        s" <li>JSON: the events will be written to the location of" +
+        s" ${ENGINE_EVENT_JSON_LOG_PATH.key}</li>" +
+        s" <li>JDBC: to be done</li>" +
+        s" <li>CUSTOM: to be done.</li></ul>")
+      .version("1.3.0")
+      .stringConf
+      .transform(_.toUpperCase(Locale.ROOT))
+      .toSequence
+      .createWithDefault(Nil)
```
Comment (ulysses-you) [inline kyuubi-common/src/main/scala/org/apache/kyuubi/config/KyuubiConf.scala:702]: we can add check value to ensure the input logger works

File: externals/kyuubi-spark-sql-engine/src/main/scala/org/apache/kyuubi/engine/spark/SparkSQLEngine.scala [hunk 8]
```diff
@@ -29,35 +29,55 @@ import org.apache.kyuubi.Logging
...
   }
 
-  override val discoveryService: Service = new EngineServiceDiscovery(this)
-
   override def initialize(conf: KyuubiConf): Unit = {
     val listener = new SparkSQLEngineListener(this)
     spark.sparkContext.addSparkListener(listener)
+    addService(eventLogging)
     super.initialize(conf)
+    eventLogging.onEvent(engineStatus.setState(ServiceState.INITIALIZED))
```
[the remark this answers is not in the document]
Comment (yaooqinn) [reply externals/kyuubi-spark-sql-engine/src/main/scala/org/apache/kyuubi/engine/spark/SparkSQLEngine.scala:55]: LGTM, by the way, copy is not safe also, we don't care much about thread-safety here

File: externals/kyuubi-spark-sql-engine/src/main/scala/org/apache/kyuubi/engine/spark/SparkSQLEngine.scala [hunk 9]
```diff
@@ -29,35 +29,55 @@ import org.apache.kyuubi.Logging
...
+    super.stop()
   }
 
   override protected def stopServer(): Unit = {
     countDownLatch.countDown()
   }
+
+  def engineId: String = {
+    (Seq(spark.sparkContext.applicationId) ++ spark.sparkContext.applicationAttemptId)
+      .mkString("_")
```
[the remark this answers is not in the document]
Comment (yaooqinn) [reply externals/kyuubi-spark-sql-engine/src/main/scala/org/apache/kyuubi/engine/spark/SparkSQLEngine.scala:79]: changed to spark.sparkContext.applicationAttemptId.getOrElse(spark.sparkContext.applicationId)

File: externals/kyuubi-spark-sql-engine/src/main/scala/org/apache/kyuubi/engine/spark/events/JsonEventLogger.scala [hunk 10]
```diff
@@ -0,0 +1,79 @@
...
+    super.stop()
+  }
+
+  override def logEvent(kyuubiEvent: KyuubiEvent): Unit = kyuubiEvent match {
+    case e: EngineEvent =>
+      val writer = getOrUpdate(e)
+      // scalastyle:off println
+      writer.println(e.toJson)
+      // scalastyle:on println
+      writer.flush()
```
[the remark this answers is not in the document]
Comment (yaooqinn) [reply externals/kyuubi-spark-sql-engine/src/main/scala/org/apache/kyuubi/engine/spark/events/JsonEventLogger.scala:76]: we can add a flush flag if needed later

File: externals/kyuubi-spark-sql-engine/src/main/scala/org/apache/kyuubi/engine/spark/events/EngineEvent.scala (hunk 2 above)
Comment (yaooqinn) [reply externals/kyuubi-spark-sql-engine/src/main/scala/org/apache/kyuubi/engine/spark/events/EngineEvent.scala:56]: moved to settings

File: externals/kyuubi-spark-sql-engine/src/main/scala/org/apache/kyuubi/engine/spark/events/EngineEvent.scala [hunk 11]
```diff
@@ -0,0 +1,117 @@
...
+ * @param shareLevel the share level for this engine
+ * @param connectionUrl the jdbc connection string
+ * @param master the master type, yarn, k8s, local etc.
+ * @param sparkVersion short version of spark distribution
+ * @param webUrl the tracking url of this engine
+ * @param startTime start time
+ * @param endTime end time
+ * @param state the engine state
+ * @param diagnostic caught exceptions if any
+ */
```
Comment (pan3793) [inline externals/kyuubi-spark-sql-engine/src/main/scala/org/apache/kyuubi/engine/spark/events/EngineEvent.scala:43]: missing `@param settings`

File: kyuubi-common/src/main/scala/org/apache/kyuubi/config/KyuubiConf.scala [hunk 12]
```diff
@@ -678,4 +678,30 @@ object KyuubiConf {
...
+
+  val ENGINE_EVENT_JSON_LOG_PATH: ConfigEntry[String] =
+    buildConf("engine.event.json.log.path")
+      .doc("The location of all the engine events go for the builtin JSON logger")
+      .version("1.3.0")
+      .stringConf
+      .createWithDefault("/tmp/kyuubi/events")
+
+  private def isValidEventLoggers(v: Set[String]): Boolean = {
+    v subsetOf(Set("SPARK", "JSON", "JDBC", "CUSTOM"))
```
Comment (pan3793) [inline kyuubi-common/src/main/scala/org/apache/kyuubi/config/KyuubiConf.scala:690]: extract `Set("SPARK", "JSON", "JDBC", "CUSTOM")` to a `val`

**Review by pan3793 — APPROVED** [review]
LGTM, left few comments.

File: externals/kyuubi-spark-sql-engine/src/main/scala/org/apache/kyuubi/engine/spark/events/EngineEvent.scala [hunk 13]
```diff
@@ -0,0 +1,147 @@
...
+    connectionUrl: String,
+    master: String,
+    deployMode: String,
+    sparkVersion: String,
+    webUrl: String,
+    driverCores: Int,
+    driverMemoryMB: Int,
+    executorCores: Int,
+    executorMemoryMB: Int,
+    maxExecutors: Int,
```
[the remark this answers is not in the document]
Comment (yaooqinn) [reply externals/kyuubi-spark-sql-engine/src/main/scala/org/apache/kyuubi/engine/spark/events/EngineEvent.scala]: it's now in setttings

**Comment by yaooqinn** [comment]
thanks, merged to master

**Pull request closed without merging**